### PR TITLE
Probabilistic Tracking Factor

### PR DESCRIPTION
**Pull request by ProfFan**
Not complete yet, need to derive the correct values used in the unit test.

### CODE REVIEW

File: Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import PenguinParallel
```
Comment (dellaert) [inline Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:1]: doc/copyright. As always

File: Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import PenguinStructures
+import TensorFlow
+
+extension AppearanceModelEncoder {
```
Comment (dellaert) [inline Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:5]: doc. What is this?

File: Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  @differentiable
+  fileprivate func encode<V: FixedSizeVector>(_ image: Tensor<Double>) -> V {
+    V(flatTensor: encode(image.expandingShape(at: 0)).squeezingShape(at: 0))
+  }
+}
+
+/// A factor over a target's pose and appearance in an image.
+public struct ProbablisticTrackingFactor<
+    Encoder: AppearanceModelEncoder,
+    ForegroundModel: GaussianModel, BackgroundModel: GaussianModel
```
Comment (dellaert) [inline Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:15]: I think these should not be GaussianModel but "GenerativeDensity"

File: Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  public var patchSize: (Int, Int)
+
+  public var appearanceModelSize: (Int, Int)
+
+  public var foregroundModel: ForegroundModel
+
+  public var backgroundModel: BackgroundModel
+  /// Creates an instance.
+  ///
+  /// - Parameters:
```
Comment (dellaert) [inline Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:39]: update parameters doc

File: Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    self.backgroundModel = backgroundModel
+  }
+
+  @differentiable
+  public func errorVector(_ pose: Pose2) -> Vector1 {
+    let region = OrientedBoundingBox(center: pose, rows: patchSize.0, cols: patchSize.1)
+    let patch = measurement.patch(at: region, outputSize: appearanceModelSize)
+    let code = encoder.encode(patch)
+
+    return Vector1(foregroundModel.negativeLogLikelihood(code) - backgroundModel.negativeLogLikelihood(code))
```
Comment (dellaert) [inline Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:68]: beautiful !!!

File: Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    self.appearanceModelSize = appearanceModelSize
+    self.foregroundModel = foregroundModel
+    self.backgroundModel = backgroundModel
+  }
+
+  @differentiable
+  public func errorVector(_ pose: Pose2) -> Vector1 {
+    let region = OrientedBoundingBox(center: pose, rows: patchSize.0, cols: patchSize.1)
+    let patch = measurement.patch(at: region, outputSize: appearanceModelSize)
+    let code = encoder.encode(patch)
```
Comment (dellaert) [inline Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:66]: I think we used `features` in the document

Comment (dellaert) [inline Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:66]: gravy: make encoder have a call function?

File: Sources/SwiftFusion/Probability/NaiveBayes.swift [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import TensorFlow
+
+
+public struct GaussianNB: GaussianModel {
```
Comment (dellaert) [inline Sources/SwiftFusion/Probability/NaiveBayes.swift:18]: doc

File: Sources/SwiftFusion/Probability/ProbabilityModel.swift [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import TensorFlow
+
+public protocol GaussianModel {
```
Comment (dellaert) [inline Sources/SwiftFusion/Probability/ProbabilityModel.swift:17]: Should be called GenerativeDensity to accommodate Histograms?

File: Tests/SwiftFusionTests/Inference/ProbablisticTrackingFactor.swift [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import TensorFlow
+import SwiftFusion
+import XCTest
+
+final class ProbablisticTrackingFactorTests: XCTestCase {
```
Comment (dellaert) [inline Tests/SwiftFusionTests/Inference/ProbablisticTrackingFactor.swift:20]: Filename should have Tests?

**Review by dellaert — APPROVED** [review]
Looks great ! Modulo comments, of course. In future document every class/function before sending a PR :-) Saves me typing...

File: Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift (hunk 2 above)
Comment (ProfFan) [reply Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:5]: removed, not relevant for the new factor

File: Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift (hunk 3 above)
Comment (ProfFan) [reply Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:15]: Changed

File: Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift (hunk 6 above)
Comment (ProfFan) [reply Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:66]: renamed

Comment (ProfFan) [reply Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:66]: the encoder can do both encode and decode, so it's kinda bad to have a call function.

I can make something with a label like `encoder(encoding: )`, but that still looks a bit weird.

File: Sources/SwiftFusion/Probability/ProbabilityModel.swift (hunk 8 above)
Comment (ProfFan) [reply Sources/SwiftFusion/Probability/ProbabilityModel.swift:17]: renamed

File: Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift (hunk 4 above)
Comment (ProfFan) [reply Sources/SwiftFusion/Inference/ProbablisticTrackingFactor.swift:46]: done

File: Tests/SwiftFusionTests/Inference/ProbablisticTrackingFactorTests.swift [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  func testErrorSanity() {
+    var v = VariableAssignments()
+    let pose = Pose2(3, 3, 0)
+    let poseId = v.store(pose)
+
+    var image: Tensor<Double> = .init(zeros: [6, 6, 3])
+
+    image[3, 3, 1] = Tensor(1.0)
+
+    var encoder = PPCA(latentSize: 1)
```
Comment (dellaert) [inline Tests/SwiftFusionTests/Inference/ProbablisticTrackingFactorTests.swift:44]: I'd assign a variable d=10 here. The fact that you can give an encoder below with d=1 and generative models with d-10 means that you might want a precondition in constructor. But hey, after this small comment just merge this PR and then make the unit test work :-)

**Pull request merged**